File: src/main/java/AoC2025_11.java
import static java.util.stream.Collectors.toMap;
import static java.util.stream.Collectors.toSet;

import com.github.pareronia.aoc.itertools.IterTools;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Stream;

@SuppressWarnings({"PMD.ClassNamingConventions", "PMD.NoPackage"})
public final class AoC2025_11 extends SolutionBase<AoC2025_11.Graph, Long, Long> {

    private static final String YOU = "you";
    private static final String SVR = "svr";
    private static final String FFT = "fft";
    private static final String DAC = "dac";
    private static final String OUT = "out";

    private AoC2025_11(final boolean debug) {
        super(debug);
    }

    public static AoC2025_11 create() {
        return new AoC2025_11(false);
    }

    public static AoC2025_11 createDebug() {
        return new AoC2025_11(true);
    }

    @Override
    protected Graph parseInput(final List<String> inputs) {
        return Graph.fromInput(inputs);
    }

    @Override
    public Long solvePart1(final Graph graph) {
        return graph.countAllPathsAlong(List.of(YOU, OUT));
    }

    @Override
    public Long solvePart2(final Graph graph) {
        return Stream.of(List.of(SVR, DAC, FFT, OUT), List.of(SVR, FFT, DAC, OUT))
                .mapToLong(graph::countAllPathsAlong)
                .sum();
    }

    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "5"),
        @Sample(method = "part2", input = TEST2, expected = "2"),
    })
    public static void main(final String[] args) throws Exception {
        create().run();
    }

    private static final String TEST1 =
            """
            aaa: you hhh
            you: bbb ccc
            bbb: ddd eee
            ccc: ddd eee fff
            ddd: ggg
            eee: out
            fff: out
            ggg: out
            hhh: ccc fff iii
            iii: out
            """;
    private static final String TEST2 =
            """
            svr: aaa bbb
            aaa: fft
            fft: ccc
            bbb: tty
            tty: ccc
            ccc: ddd eee
            ddd: hub
            hub: fff
            eee: dac
            dac: fff
            fff: ggg hhh
            ggg: out
            hhh: out
            """;

    record Graph(Map<String, Set<String>> edges) {

        public static Graph fromInput(final List<String> inputs) {
            final Map<String, Set<String>> edges =
                    inputs.stream()
                            .collect(
                                    toMap(
                                            line -> line.split(": ")[0],
                                            line ->
                                                    Arrays.stream(line.split(": ")[1].split(" "))
                                                            .collect(toSet())));
            return new Graph(edges);
        }

        private long countAllPaths(final String start, final String end) {
            class DFS {
                private final Map<String, Long> cache;

                protected DFS() {
                    this.cache = new HashMap<>();
                }

                @SuppressWarnings("PMD.OnlyOneReturn")
                public long dfs(final String node) {
                    if (this.cache.containsKey(node)) {
                        return this.cache.get(node);
                    }
                    if (end.equals(node)) {
                        return 1;
                    }
                    final long ans =
                            Graph.this.edges.getOrDefault(node, Set.of()).stream()
                                    .mapToLong(this::dfs)
                                    .sum();
                    this.cache.put(node, ans);
                    return ans;
                }
            }

            return new DFS().dfs(start);
        }

        public long countAllPathsAlong(final List<String> along) {
            return IterTools.pairwise(along).stream()
                    .mapToLong(pair -> this.countAllPaths(pair.first(), pair.second()))
                    .reduce((acc, cnt) -> acc * cnt)
                    .getAsLong();
        }
    }
}

File: src/main/python/AoC2025_11.py
#! /usr/bin/env python3
#
# Advent of Code 2025 Day 11
#

import itertools
import sys
from dataclasses import dataclass
from functools import cache
from math import prod
from pathlib import Path
from typing import Self

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

TEST1 = """\
aaa: you hhh
you: bbb ccc
bbb: ddd eee
ccc: ddd eee fff
ddd: ggg
eee: out
fff: out
ggg: out
hhh: ccc fff iii
iii: out
"""
TEST2 = """\
svr: aaa bbb
aaa: fft
fft: ccc
bbb: tty
tty: ccc
ccc: ddd eee
ddd: hub
hub: fff
eee: dac
dac: fff
fff: ggg hhh
ggg: out
hhh: out
"""


@dataclass(frozen=True)
class Graph:
    edges: dict[str, set[str]]

    @classmethod
    def from_input(cls, inputs: InputData) -> Self:
        return cls(
            {
                from_[:-1]: set(to)
                for from_, *to in (line.split() for line in inputs)
            }
        )

    def count_all_paths_along(self, along: list[str]) -> int:
        def count_all_paths(start: str, end: str) -> int:
            @cache
            def dfs(node: str) -> int:
                if node == end:
                    return 1
                return sum(dfs(nxt) for nxt in self.edges.get(node, set()))

            return dfs(start)

        return prod(
            count_all_paths(n1, n2) for n1, n2 in itertools.pairwise(along)
        )

    def output_graph(self) -> None:
        with Path("~/tmp/2025_11.dot").open("w") as f:
            f.write("digraph G {\n")
            f.write('fft [style="filled" fillcolor=red]\n')
            f.write('dac [style="filled" fillcolor=red]\n')
            for k, vv in self.edges.items():
                for v in vv:  # noqa:FURB122
                    f.write(f"{k} -> {v}\n")
            f.write("}\n")


Input = Graph
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return Graph.from_input(input_data)

    def part_1(self, graph: Graph) -> Output1:
        return graph.count_all_paths_along(["you", "out"])

    def part_2(self, graph: Graph) -> Output2:
        return graph.count_all_paths_along(
            ["svr", "dac", "fft", "out"]
        ) + graph.count_all_paths_along(["svr", "fft", "dac", "out"])

    @aoc_samples(
        (
            ("part_1", TEST1, 5),
            ("part_2", TEST2, 2),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2025, 11)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
